### ISSUE TITLE
Add 'T' and 'L' as new optional coordinates in WKT literals

### ISSUE THREAD

**Issue by mjovanovik** [opengeospatial/ogc-geosparql#35]
We propose an enhancement of the modifiers available in WKT literals with two additional optional coordinates, named _T_ and _L_, that could be used in the following manner:

- _T_ should be used for the temporal aspect of the geometry, i.e. a timestamp of a geometry denoting an additional dimension.
- _L_ should be used for specifying the level of detail (zoom level, granularity or scale), as an additional dimension.

The rationale behind _T_ is to have the ability to add a time dimension to a geometry, which can be used in various scenarios, such as defining the time the geometry came into existence. This is especially important when dealing with traffic data on a map, for instance, where a vehicle position has a time dimension as well.

The rationale behind _L_ is to have the ability to determine the importance of a geometry when displaying it on a map, or returning it as part of a SELECT query. One way to do this is to denote the "level of detail" for a geometry, as an integer, which will be used in the calculations which determine whether the geometry should be selected for the current scale, or not.

These two optional modifiers should be treated in the same way as _Z_ and _M_, thus any geometry could have them specified. For example, a point could be represented from 2D to 6D, all depending on which modifiers are present:

- 2D: 	POINT(X Y)
- 3D: 	POINTZ(X Y Z), POINTM(X Y M), POINTT(X Y T), POINTL(X Y L)
- 4D: 	POINTZM(X Y Z M), POINTZT(X Y Z T), POINTZL(X Y Z L), POINTMT(X Y M T), POINTML(X Y M L), POINTTL(X Y T L)
- 5D: 	POINTZMT(X Y Z M T), POINTZML(X Y Z M L), POINTZTL(X Y Z T L), POINTMTL(X Y M T L) 
- 6D:	POINTZMTL(X Y Z M T L)

This also applies to other geometries (LINESTRING, POLYGON, etc...) in a similar way.

### Examples

**Example 1**. If the _T_ modifier is introduced, we can use a POINTT type to add a timestamp for a geometry directly in the WKT literal:
```
<#p0> geo:hasGeometry <#p0_g>.
<#p0_g> geo:asWKT "POINTT(12.27 51.32 2017-09-03T12:48:11Z)"^^geo:wktLiteral.
```
instead of using a POINT type with an additional property (`hasTimestamp`) for time. 
```
<#p0> geo:hasGeometry <#p0_g>.
<#p0_g> geo:asWKT "POINT(12.27 51.32)"^^geo:wktLiteral.
<#p0> :hasTimestamp "2017-09-03T12:48:11Z"^^xsd:dateTime.
```

**Example 2**. We can have a similar situation with a 3D geometry which needs to have a time dimension as well, with the POINTZT type:
```
<#p0> geo:hasGeometry <#p0_g>.
<#p0_g> geo:asWKT "POINTZT(12.27 51.32 120 
2017-09-03T12:48:11Z)"^^geo:wktLiteral.
```
instead of using a POINTZ type with an additional property (`hasTimestamp`) for time. 
```
<#p0> geo:hasGeometry <#p0_g>.
<#p0_g> geo:asWKT "POINTZ(12.27 51.32 120)"^^geo:wktLiteral.
<#p0> :hasTimestamp "2017-09-03T12:48:11Z"^^xsd:dateTime.
```

**Example 3**. We can combine the time and level of detail information within a WKT literal using the POINTZTL type:
```
<#p0> geo:hasGeometry <#p0_g>.
<#p0_g> geo:asWKT "POINTZTL(12.27 51.32 120 
2017-09-03T12:48:11Z 6)"^^geo:wktLiteral.
```
instead of using a POINTZ type with additional properties (`hasTimestamp`, `scale`) for time and level of detail, respectively.
```
<#p0> geo:hasGeometry <#p0_g>.
<#p0_g> geo:asWKT "POINTZ(12.27 51.32 120)"^^geo:wktLiteral.
<#p0> :hasTimestamp "2017-09-03T12:48:11Z"^^xsd:dateTime.
<#p0> :scale "6"^^xsd:integer.
```

### Background

Our team at [OpenLink Software](https://www.openlinksw.com) has been working on 5D geospatial support in its [Virtuoso](https://virtuoso.openlinksw.com) product, as part of the [SAGE: Semantic Geospatial Analytics](http://sageproject.eu/) project.

As part of our work, we introduced and successfully used these WKT and GeoSPARQL enhancements, as we outlined in our Deliverable 4.2 "[Prototype verifying 5D Geospatial support in Virtuoso](https://drive.google.com/file/d/1EF7FQrIP4VzkNTLd-Cc2eh88SFyQHQ49/view?usp=sharing)".

**Comment by jabhay** [comment]
Created MoSCoW poll

[![](https://api.gh-polls.com/poll/01EJRE8MEJ2GMYMQQG6MKVAR4N/Must)](https://api.gh-polls.com/poll/01EJRE8MEJ2GMYMQQG6MKVAR4N/Must/vote)
[![](https://api.gh-polls.com/poll/01EJRE8MEJ2GMYMQQG6MKVAR4N/Should)](https://api.gh-polls.com/poll/01EJRE8MEJ2GMYMQQG6MKVAR4N/Should/vote)
[![](https://api.gh-polls.com/poll/01EJRE8MEJ2GMYMQQG6MKVAR4N/Could)](https://api.gh-polls.com/poll/01EJRE8MEJ2GMYMQQG6MKVAR4N/Could/vote)
[![](https://api.gh-polls.com/poll/01EJRE8MEJ2GMYMQQG6MKVAR4N/Won%27t)](https://api.gh-polls.com/poll/01EJRE8MEJ2GMYMQQG6MKVAR4N/Won%27t/vote)

**Comment by FransKnibbe** [comment]
I have just voted W, which I'd like to explain:

- In general, I think WKT literals in the current GeoSPARQL spec are overloaded as they are. I would rather see them slimmed down (i.e. remove the CRS bit) than expanded further.
- I think that for most knowledge domains (excluding cosmology and quantum physics), space and time can be considered as being independent. So I think it best to keep ontologies for space and time nicely separated.
- Of course it should be possible to use time stamps with geometry, but I think it would be better to use separate properties for that purpose, which would also make clear what exactly is timed. For instance, a set of coordinated could be measured in the field and be transformed to another CRS at different times.
- A way to specify spatial resolution is absolutely required, but I think it would be much better to model that as a separate property of geometry.

To summarize: yes to improving expressing time and LoD, no to adding them to the WKT literal.

**Comment by paulc-dstl** [comment]
I agree with @FransKnibbe - "_So I think it best to keep ontologies for space and time nicely separated._" The Owl-Time ontology, for example, can be seen as largely analogous to GeoSPARQL in the way it works. And if we are talking about temporal dimensions of spatial things, I would suggest we are actually talking about things which have both spatial and temporal dimensions (eg a forest fire, an observation of a car in traffic, the Bronze Age, the Great Fire of London, etc), ie there is some class of things which exhibit both spatial and temporal characteristics, the former can be modelled using a geospatial ontology and the latter by a temporal ontology, both of which can be nested within an appropriate domain ontology for the thing. A simple timestamp attached to a geo thing would be far too crude for many purposes and would get in the way of richer, more descriptive approaches to temporal things (instants, intervals, etc) and relationships between them (ie Allen's temporal algebra). Let's keep GeoSPARQL simple and focussed.

**Comment by nicholascar** [comment]
GeoSAPRQL 1.1 as published states:

> GeoSPARQL does not directly provide support for temporality. Predicates for temporal relations may be used from the OWL Time Ontology [TIME](https://docs.ogc.org/is/22-047r1/22-047r1.html#TIME), but query extension functions for spatiotemporal operations are not present in the GeoSPARQL standard.

So I'm closing this issue now as the answer is "wontfix", at least for 1.1.
